<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Driven by biosustain</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-light.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <!-- highlight.js -->
    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/8.8.0/styles/googlecode.min.css">
    <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/8.8.0/highlight.min.js"></script>


  </head>
  <body>
    <div class="wrapper">
      <header>
          <img src="driven_logo.png"/>
          <p style="text-align: center;"><strong style="color: red;">Site under construction...</strong></p>
        <ul>
          <li><a href="https://github.com/biosustain/driven/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/biosustain/driven/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/biosustain/driven">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h3>
<a id="welcome-to-driven" class="anchor" href="#welcome-to-driven" aria-hidden="true"><span class="octicon octicon-link"></span></a>Welcome to Driven.</h3>

<p>Driven is a <i>python</i> package providing methods to analyze <i>omics</i> data using genome-scale models.</p>

<h3>
<a id="description" class="anchor" href="#description" aria-hidden="true"><span class="octicon octicon-link"></span></a>Description</h3>

<p>Driven provides <a href="#state-of-the-art-methods">state-of-the-art methods</a> to analyze <i>omics</i> using constraint-based modeling. State-of-the-art methods have been implemented as well as new methods. All of them are accessible via a friendly <a>API</a>.</p>
<p>Driven is integrated with several libraries: <a href="https://opencobra.github.io/cobrapy/" >cobrapy</a> for easy access to Constraint-based Models, <a href="http://escher.github.io">ESCHER</a> it allows easy visualization of the results.
All of this can be used within jupyter notebooks.</p>

<h3>
<a id="state-of-the-art-methods" class="anchor" href="#state-of-the-art-methods" aria-hidden="true"><span class="octicon octicon-link"></span></a>State-of-the-art Methods</h3>

    <h4 class="methods-group"><i>Transcriptomics</i></h4>
    <ul>
        <li>GIMME <code>driven.flux_analysis.transcriptomics.gimme</code></li>
        <li>iMAT <code>driven.flux_analysis.transcriptomics.imat</code></li>
    </ul>
    <h4 class="methods-group"><i>Proteomics</i></h4>
    <ul>
        <li>GIMME <code>driven.flux_analysis.proteomics.gimme</code></li>
    </ul>
    <h4 class="methods-group"><i>Thermodynamics</i></h4>
    <ul>
        <li>TFBA <code>driven.flux_analysis.thermodynamics.tfba</code></li>
    </ul>
    <h4 class="methods-group"><i>Fluxomics</i></h4>
    <ul>
        <li>FBA <code>driven.flux_analysis.fluxomics.fba</code></li>
    </ul>



<h3>
<a id="easy-access" class="anchor" href="#easy-access" aria-hidden="true"><span class="octicon octicon-link"></span></a>Easy data access</h3>
    Importing data into driven is easy. Driven reads tabular files directly:
    <h4 class="methods-group"><i>Expression Profiles</i></h4>
          Expression data can be easily imported from comma separated values (CSV) files:
          <pre><code class="python">from driven.datasets.expression_profile import ExpressionProfile
data = ExpressionProfile.from_csv("your_data.csv")</code></pre>
          If the data in the file needs to be filtered or processed, driven can import data from pandas <a href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html">DataFrame</a>.
          <pre><code class="python">data = ExpressionProfile.from_data_frame(dataframe)</code></pre>

    <h4 class="methods-group"><i>Flux data</i></h4>
          It is also very simple to compute a flux distribution using flux measurements, e.g. 13C fluxes.
          The flux constraints can be imported in two formats: measurements +- error, or upper and lower limits.

          <pre><code class="python">from driven.datasets.fluxes import FluxConstraints
# measurements
flux_constraints = FluxConstraints.from_csv("measurments.csv", type="measurements")
# constraints
flux_constraints = FluxConstraints.from_csv("constraints.csv", type="constraints")</code></pre>

    More details about pre-processing data sets can be found in <a href="">Workflows</a>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/biosustain">biosustain</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
     <script>hljs.initHighlightingOnLoad();</script>
    <script src="javascripts/scale.fix.js"></script>
    
  </body>
</html>
